contracts/crosschain/Crosschain.sol

```diff
@@ -0,0 +1,147 @@
+
+        // Ensure data and nonce provided.
+        require(state.data != bytes32(0), "CrosschainTransfer: EMPTY_DATA");
+        require(state.nonce != uint256(0), "CrosschainTransfer: EMPTY_NONCE");
+
+        // Initiator balance must be greater than 0 and include amount for fee.
+        require(
+            balance.amount[0] > 0,
+            "CrosschainTransfer: ZER0_SENDER_BALANCE"
+        );
+        require(
+            state.fee <= balance.amount[0],
+            "CrosschainTransfer: INSUFFICIENT_BALANCE"
+        );
+
+        // Recipient balance must be 0.
+        require(
+            balance.amount[1] == 0,
+            "CrosschainTransfer: NONZERO_RECIPIENT_BALANCE"
+        );
+
+        // Valid lockHash to secure funds must be provided.
+        require(
+            state.lockHash != bytes32(0),
+            "CrosschainTransfer: EMPTY_LOCKHASH"
+        );
+
+        // Update state.
+        balance.amount[1] = balance.amount[0];
+        balance.amount[0] = 0;
+
+        // Valid initial transfer state
+        return true;
+    }
+
+    function resolve(
+        bytes calldata encodedBalance,
+        bytes calldata encodedState,
+        bytes calldata encodedResolver
+    ) external pure override returns (Balance memory) {
+        TransferState memory state = abi.decode(encodedState, (TransferState));
+        TransferResolver memory resolver =
+            abi.decode(encodedResolver, (TransferResolver));
+        Balance memory balance = abi.decode(encodedBalance, (Balance));
+
+        // Ensure data and nonce provided.
+        require(state.data != bytes32(0), "CrosschainTransfer: EMPTY_DATA");
+        require(state.nonce != uint256(0), "CrosschainTransfer: EMPTY_NONCE");
+
+        // Amount recipient is able to withdraw > 0.
+        require(
+            balance.amount[1] == 0,
+            "CrosschainTransfer: NONZERO_RECIPIENT_BALANCE"
+        );
+
+        // Transfer must have two valid parties.
+        require(
+            state.initiator != address(0) && state.responder != address(0),
+            "CrosschainTransfer: EMPTY_SIGNERS"
+        );
+
+        // Both signatures must be valid.
+        require(
+            state.data.checkSignature(
+                resolver.initiatorSignature,
+                state.initiator
+            ),
+            "CrosschainTransfer: INVALID_INITIATOR_SIG"
+        );
+        require(
+            state.data.checkSignature(
+                resolver.responderSignature,
+                state.responder
+            ),
+            "CrosschainTransfer: INVALID_RESPONDER_SIG"
+        );
+
+        // Check hash for normal payment unlock.
+        bytes32 generatedHash = sha256(abi.encode(resolver.preImage));
+        require(
+            state.lockHash == generatedHash,
+            "CrosschainTransfer: INVALID_PREIMAGE"
+        );
+
+        // Reduce CrosschainTransfer amount by optional fee.
```

Comment:
"Reduce CrosschainTransfer amount by optional fee" -> "Reduce CrosschainTransfer amount to optional fee"